### PR TITLE
Remove references to `pkg/apis/componentconfig/install` and fix inconsistencies

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:
This PR fixes a bunch of problems with the (deprecated, monolithic, to be removed) componentconfig API group:
 - As discussed with @liggitt, the current structure of k-c-m componentconfig types is really bad, and we don't want anyone to think it can be serialized (not that we have any such code, but anyway). So we deciided to remove all JSON tags from the types to be consistent, register the violations, then move types out, get the structure right, and **first then** add JSON tags.
 - There should not be an `install` package for componentconfigs that installs stuff in `legacyscheme`,  removed the package and references to it.
 - Added myself and @sttts to approvers for `pkg/apis/componentconfig`, which we intend to remove in favor for dedicated API groups in different staging repos.
 - Removed the componentconfig types from the API testing in `pkg/api/testing`, that roundtrip and defaulting testing will exist in the dedicated API groups instead. Added a TODO to add roundtrip, defaulting and type tag testing later.
 - Made the `register.go` files more consistent to the common template.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
ref https://github.com/kubernetes/community/pull/2354

**Special notes for your reviewer**:

This PR must merge before https://github.com/kubernetes/kubernetes/pull/66722 and https://github.com/kubernetes/kubernetes/pull/66993

**Release note**:

```release-note
NONE
```
@kubernetes/sig-api-machinery-pr-reviews @kubernetes/sig-cluster-lifecycle-pr-reviews 
/assign @sttts @thockin @jbeda @liggitt

### CODE REVIEW

File: pkg/apis/componentconfig/v1alpha1/defaults.go [hunk 1]
```diff
@@ -45,7 +45,7 @@ func SetDefaults_CloudControllerManagerConfiguration(obj *CloudControllerManager
 
 func SetDefaults_KubeControllerManagerConfiguration(obj *KubeControllerManagerConfiguration) {
 	zero := metav1.Duration{}
-	if len(obj.Controllers) == 0 {
+	if obj.Controllers == nil || len(obj.Controllers) == 0 {
```
Comment (liggitt) [inline pkg/apis/componentconfig/v1alpha1/defaults.go]: a nil slice is also zero-length... this shouldn't be needed

File: pkg/apis/componentconfig/v1alpha1/types.go [hunk 2]
```diff
@@ -21,16 +21,18 @@ import (
 )
 
 // ClientConnectionConfiguration contains details for constructing a client.
+// TODO: This struct should be referenced from k8s.io/apimachinery/pkg/apis/config instead.
 type ClientConnectionConfiguration struct {
 	// kubeConfigFile is the path to a kubeconfig file.
-	KubeConfigFile string `json:"kubeconfig"`
+	// TODO: This field maybe should be named just `KubeConfig`
+	KubeConfigFile string `json:"kubeConfigFile"`
```
Comment (liggitt) [inline pkg/apis/componentconfig/v1alpha1/types.go]: is there a reason you're changing the external serialization of this? I expected this instead:
```
	Kubeconfig string `json:"kubeconfig"`
```

File: pkg/apis/componentconfig/v1alpha1/types.go [hunk 3]
```diff
@@ -210,70 +212,70 @@ type KubeControllerManagerConfiguration struct {
...
-	JobController JobControllerConfiguration
+	JobController JobControllerConfiguration `json:"jobController"`
 	// NamespaceControllerConfiguration holds configuration for NamespaceController
 	// related features.
-	NamespaceController NamespaceControllerConfiguration
+	NamespaceController NamespaceControllerConfiguration `json:"namespaceController"`
 	// NodeIpamControllerConfiguration holds configuration for NodeIpamController
 	// related features.
-	NodeIpamController NodeIpamControllerConfiguration
+	NodeIpamController NodeIpamControllerConfiguration `json:"nodeIpamController"`
```
Comment (liggitt) [inline pkg/apis/componentconfig/v1alpha1/types.go]: `Ipam` or `IPAM`?

File: pkg/apis/componentconfig/v1alpha1/types.go [hunk 4]
```diff
@@ -210,70 +212,70 @@ type KubeControllerManagerConfiguration struct {
...
-	ReplicationController ReplicationControllerConfiguration
+	ReplicationController ReplicationControllerConfiguration `json:"replicationController"`
 	// ResourceQuotaControllerConfiguration holds configuration for
 	// ResourceQuotaController related features.
-	ResourceQuotaController ResourceQuotaControllerConfiguration
+	ResourceQuotaController ResourceQuotaControllerConfiguration `json:"resourceQuotaController"`
 	// SAControllerConfiguration holds configuration for ServiceAccountController
 	// related features.
-	SAController SAControllerConfiguration
+	SAController SAControllerConfiguration `json:"saController"`
```
Comment (liggitt) [inline pkg/apis/componentconfig/v1alpha1/types.go]: `sa` or `serviceAccount`?

File: pkg/apis/componentconfig/v1alpha1/types.go [hunk 5]
```diff
@@ -292,25 +294,25 @@ type CloudControllerManagerConfiguration struct {
...
-	ServiceController ServiceControllerConfiguration
+	ServiceController ServiceControllerConfiguration `json:"serviceController"`
 	// NodeStatusUpdateFrequency is the frequency at which the controller updates nodes' status
-	NodeStatusUpdateFrequency metav1.Duration
+	NodeStatusUpdateFrequency metav1.Duration `json:"nodeStatusUpdateFrequency"`
 }
 
 type CloudProviderConfiguration struct {
 	// Name is the provider for cloud services.
-	Name string `json:"cloudProvider"`
```
Comment (liggitt) [inline pkg/apis/componentconfig/v1alpha1/types.go]: same question, why change external serialization instead of changing the field name?

File: pkg/apis/componentconfig/v1alpha1/types.go (hunk 4 above)
Comment (luxas) [reply pkg/apis/componentconfig/v1alpha1/types.go]: probably `serviceAccount` after a refactor, but this just makes things consistent without renaming fields as a stop-gap. I want to keep this PR as small as possible and do field rename refactor in https://github.com/kubernetes/kubernetes/pull/66993

File: pkg/apis/componentconfig/v1alpha1/types.go (hunk 5 above)
Comment (luxas) [reply pkg/apis/componentconfig/v1alpha1/types.go]: I plan to fix field naming and structure later in https://github.com/kubernetes/kubernetes/pull/66993, but now just make stuff consistent to make it pass validation

File: pkg/apis/componentconfig/v1alpha1/defaults.go (hunk 1 above)
Comment (luxas) [reply pkg/apis/componentconfig/v1alpha1/defaults.go]: ok, right, removing then

File: pkg/apis/componentconfig/v1alpha1/types.go (hunk 5 above)
Comment (liggitt) [reply pkg/apis/componentconfig/v1alpha1/types.go]: is the assumption that none of these serialized fields are currently used in supported config files? changing serialization tags is not something we should generally do. I'd rather see this validation errors fixed by changing the go field name if possible.

File: pkg/apis/componentconfig/v1alpha1/types.go (hunk 3 above)
Comment (luxas) [reply pkg/apis/componentconfig/v1alpha1/types.go]: IPAM indeed, but later

File: pkg/apis/componentconfig/v1alpha1/types.go (hunk 4 above)
Comment (liggitt) [reply pkg/apis/componentconfig/v1alpha1/types.go]: Is the assumption that none of these serialized fields are currently used in supported config files? It's very clear this struct should not be serialized, since it is missing serialization tags. Once you add them in this PR, it's a lot harder to tell whether someone has ever written out config using this struct they expect to be able to read back in. Can we correct the field names first before adding tags for serialization?

File: pkg/apis/componentconfig/v1alpha1/types.go (hunk 3 above)
Comment (liggitt) [reply pkg/apis/componentconfig/v1alpha1/types.go]: same comment in https://github.com/kubernetes/kubernetes/pull/67090#discussion_r208331956 ... let's not add tags for serialization we intend to change

**Comment by luxas** [comment]
Commenting top-level here @liggitt; I intend to keep this PR minimal, with no struct / field changes so it can merge ASAP to unblock the other PRs I have open: https://github.com/kubernetes/kubernetes/pull/66993 and https://github.com/kubernetes/kubernetes/pull/66722.
https://github.com/kubernetes/kubernetes/pull/66722's scope is to remove the redundant structs and start using shared `k8s.io/{apimachinery,apiserver}/pkg/apis/config` structs in `pkg/apis/componentconfig`.
https://github.com/kubernetes/kubernetes/pull/66993's scope is to rename the fields, and actually make the structure of the fields and naming somewhat sane before moving `KubeSchedulerConfiguration` and `KubeControllerManagerConfiguration` out to different API groups/repos.

TL;DR; I don't want to rename fields here in this PR as it starts touching the whole codebase instantly.
And these structs aren't exposed to the end user in any way, the current componentconfig API group is unfunctional as it stands, and only used internally, but neither `kube-scheduler` or `kube-controller-manager` supports reading a config file properly really yet, so we have full freedom to move stuff around and fix stuff now.

**Comment by luxas** [comment]
> Can we correct the field names first before adding tags for serialization?

We haven't decided a structure yet, and that is ultimately up to the k-c-m OWNERS to do before we release v1.12 with the types in the dedicated API groups. Right now we have to do this incrementally in order to move stuff out successfully. Also, pragmatically there's no chance we'd be able to change this structure in one PR, as we don't know yet what the end result will look like. Right now this is an internal-only API that we're about to expose, and it will take time to find the right structure. I hope it's ok for you to do this step-by-step like this.

File: cmd/kube-scheduler/app/options/options.go [hunk 6]
```diff
@@ -156,8 +157,10 @@ func (o *Options) ApplyTo(c *schedulerappconfig.Config) error {
 			return err
 		}
 	} else {
+		fmt.Println("load config", o.ConfigFile)
```
Comment (liggitt) [inline cmd/kube-scheduler/app/options/options.go]: remove debug

File: cmd/kube-scheduler/app/options/options_test.go [hunk 7]
```diff
@@ -35,7 +35,7 @@ func TestSchedulerOptions(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	defer os.RemoveAll(tmpDir)
+	//defer os.RemoveAll(tmpDir)
```
Comment (liggitt) [inline cmd/kube-scheduler/app/options/options_test.go]: remove debug?

File: cmd/kube-scheduler/app/options/options_test.go [hunk 8]
```diff
@@ -67,7 +67,7 @@ func TestSchedulerOptions(t *testing.T) {
 apiVersion: componentconfig/v1alpha1
 kind: KubeSchedulerConfiguration
 clientConnection:
-  kubeconfig: "%s"
+  kubeConfigFile: "%s"
```
Comment (liggitt) [inline cmd/kube-scheduler/app/options/options_test.go]: revert this

**Comment by liggitt** [comment]
> Right now we have to do this incrementally in order to move stuff out successfully.

Incremental is fine, but let's not add serialization tags we know we don't want as an intermediate step... it makes it really hard to track what was previously serialized this way in a released version.

> Right now this is an internal-only API that we're about to expose, and it will take time to find the right structure.

Not all the structs you are touching are internal only. kube-scheduler (which is affected by the kubeConfigFile change in this PR) has been telling people to run from config files [since 1.9](https://github.com/kubernetes/kubernetes/blob/release-1.9/plugin/cmd/kube-scheduler/app/server.go#L107)

> I hope it's ok for you to do this step-by-step like this.

step by step is fine, but I'd start by moving things into place first, rather than making things serializable prematurely.

File: pkg/apis/componentconfig/scheme/scheme.go [hunk 9]
```diff
@@ -14,24 +14,31 @@ See the License for the specific language governing permissions and
...
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
-	"k8s.io/kubernetes/pkg/api/legacyscheme"
 	"k8s.io/kubernetes/pkg/apis/componentconfig"
 	"k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1"
 )
 
+// Scheme is the runtime.Scheme to which all kubeadm api types are registered.
+var Scheme = runtime.NewScheme()
```
Comment (sttts) [inline pkg/apis/componentconfig/scheme/scheme.go]: I don't think we want this global scheme. This is the wrong direction.

File: cmd/kube-scheduler/app/options/configfile.go [hunk 10]
```diff
@@ -18,28 +18,16 @@ package options
...
-	"k8s.io/apimachinery/pkg/runtime/serializer"
 	"k8s.io/apimachinery/pkg/runtime/serializer/json"
-	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/kubernetes/pkg/apis/componentconfig"
+	componentconfigscheme "k8s.io/kubernetes/pkg/apis/componentconfig/scheme"
 	componentconfigv1alpha1 "k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1"
 )
 
-var (
-	configScheme = runtime.NewScheme()
```
Comment (sttts) [inline cmd/kube-scheduler/app/options/configfile.go]: why remove this local scheme? This was the better approach.

File: pkg/apis/componentconfig/scheme/scheme.go (hunk 9 above)
Comment (luxas) [reply pkg/apis/componentconfig/scheme/scheme.go]: Right, we don't want this for the monolithic `componentconfig` API group, but we want it next to the individual API groups. This PR makes `componentconfig` follow that pattern, so it's easy to just move the reference later.

File: cmd/kube-scheduler/app/options/configfile.go (hunk 10 above)
Comment (luxas) [reply cmd/kube-scheduler/app/options/configfile.go]: See the other comment

File: cmd/kube-scheduler/app/options/options_test.go (hunk 7 above)
Comment (luxas) [reply cmd/kube-scheduler/app/options/options_test.go]: right, fixed

File: cmd/kube-scheduler/app/options/options.go (hunk 6 above)
Comment (luxas) [reply cmd/kube-scheduler/app/options/options.go]: fixed

File: pkg/api/testing/serialization_test.go [hunk 11]
```diff
@@ -205,11 +205,7 @@ func TestCommonKindsRegistered(t *testing.T) {
 func TestRoundTripTypes(t *testing.T) {
 	seed := rand.Int63()
 	fuzzer := fuzzer.FuzzerFor(FuzzerFuncs, rand.NewSource(seed), legacyscheme.Codecs)
-
-	nonRoundTrippableTypes := map[schema.GroupVersionKind]bool{
-		{Group: "componentconfig", Version: runtime.APIVersionInternal, Kind: "KubeProxyConfiguration"}:     true,
-		{Group: "componentconfig", Version: runtime.APIVersionInternal, Kind: "KubeSchedulerConfiguration"}: true,
-	}
+	nonRoundTrippableTypes := map[schema.GroupVersionKind]bool{}
```
Comment (sttts) [inline pkg/api/testing/serialization_test.go]: do we have a roundtrip_test.go like https://github.com/kubernetes/kubernetes/blob/d3bd0eb1d5cefc25e4476d8dc086ebd90439ef4e/staging/src/k8s.io/apiserver/pkg/apis/audit/install/roundtrip_test.go#L26 for componentconfig types? If not, that would be a good follow-up.

File: pkg/apis/componentconfig/install/install.go [hunk 12]
```diff
@@ -1,38 +0,0 @@
...
-import (
-	"k8s.io/apimachinery/pkg/runtime"
-	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
-	"k8s.io/kubernetes/pkg/api/legacyscheme"
-	"k8s.io/kubernetes/pkg/apis/componentconfig"
-	"k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1"
-)
-
-func init() {
-	Install(legacyscheme.Scheme)
```
Comment (sttts) [inline pkg/apis/componentconfig/install/install.go]: :+1:

File: hack/verify-api-groups.sh [hunk 13]
```diff
@@ -94,6 +94,7 @@ done
 packages_without_install=(
 	"k8s.io/kubernetes/pkg/apis/abac"
 	"k8s.io/kubernetes/pkg/apis/admission"
+	"k8s.io/kubernetes/pkg/apis/componentconfig" # TODO: Remove this package completely and from this list
```
Comment (sttts) [inline hack/verify-api-groups.sh]: under which conditions?

**Comment by sttts** [comment]
Just some nits. Lgtm otherwise.

File: hack/verify-api-groups.sh (hunk 13 above)
Comment (luxas) [reply hack/verify-api-groups.sh]: when the directory is removed we can get rid of this. Is the comment clear enough?

File: pkg/api/testing/serialization_test.go (hunk 11 above)
Comment (luxas) [reply pkg/api/testing/serialization_test.go]: I added a TODO for this in fuzzer.go in the componentconfig pkg, but will be applied first when the structs are split out to different repos.

File: hack/verify-api-groups.sh (hunk 13 above)
Comment (sttts) [reply hack/verify-api-groups.sh]: missed the "and".  it's ok.

File: pkg/apis/componentconfig/v1alpha1/types.go [hunk 14]
```diff
@@ -152,28 +153,28 @@ type KubeSchedulerLeaderElectionConfiguration struct {
 type PersistentVolumeRecyclerConfiguration struct {
 	// maximumRetry is number of retries the PV recycler will execute on failure to recycle
 	// PV.
-	MaximumRetry int32 `json:"maximumRetry"`
+	MaximumRetry int32
```
Comment (sttts) [inline pkg/apis/componentconfig/v1alpha1/types.go]: why remove the tags?

Comment (luxas) [reply pkg/apis/componentconfig/v1alpha1/types.go]: see @liggitt's and my conversation in Slack and the PR description

Comment (liggitt) [reply pkg/apis/componentconfig/v1alpha1/types.go]: the k-c-m structs had a mix of missing tags, incorrect tags, and correct tags, but it didn't really matter because those structs were not currently used for deserializing anything yet.

this tries to make it clear which structs are currently being used for serialization and which are free to be changed/corrected without regard for existing deserialization use

for k-c-m, we need to:
1. move structs to the right packages
2. pick the actual field names and structure to use, then add serialization tags (resulting in zero violations found by the verifier in the process)
4. add deserialization paths

**Comment by liggitt** [comment]
/approve

@sttts has lgtm

**Comment by sttts** [comment]
needs rebase

**Comment by luxas** [comment]
@sttts rebased

**Comment by luxas** [comment]
LGTM'ing with @sttts approval
